scripts/jenkins/ardana/gerrit/gerrit.py

```diff
@@ -0,0 +1,280 @@
+        direct and indirect dependencies.
+
+        :return collected GerritChange objects
+        """
+
+        if self._dependencies is not None:
+            return self._dependencies
+
+        self._dependencies = [self] + self._get_submitted_together()
+        for change in dependencies:
+            change._load_dependencies(self._dependencies)
+        self._dependencies.pop(0)
+        return self._dependencies
+
+    def depends_on_directly(self, change):
+        """
+        Check if the supplied GerritChange object is a direct dependency
+        of this change, specified through a Depends-On commit message marker
+        or submitted together with it.
+
+        :param change: GerritChange object
+        :return: bool
+        """
+        deps = self.find_dependency_headers()
+        submitted_together = self._get_submitted_together()[:1]
+        return change.id in deps or change.change_id in deps or \
+               change == submitted_together
+
+    def review(self, label=None, vote=1, message=''):
+        print("Posting {} review for change: {}".format(
+            " {}: {}".format(label, vote) if label else '', self))
+        auth = HTTPDigestAuthFromNetrc(url=GERRIT_URL)
+        rest = GerritRestAPI(url=GERRIT_URL, auth=auth, verify=GERRIT_VERIFY)
+        rev = GerritReview()
+        rev.set_message(message)
+        if label:
+            rev.add_labels({label: vote})
+
+        rest.review(self.id, self.patchset, rev)
+
+    def merge(self):
+        auth = HTTPDigestAuthFromNetrc(url=GERRIT_URL)
+        rest = GerritRestAPI(url=GERRIT_URL, auth=auth, verify=GERRIT_VERIFY)
+        url_path = '/changes/{}/submit'.format(self.id)
+        rest.post(url_path)
+
+    def __repr__(self):
+        return "<GerritChange {}/{} ({}/{}): '{}'>".format(
+            self.id, self.patchset, self.gerrit_project,
+            self.branch, self.subject)
+
+
+class GerritChangeSet:
+    """
+    Holds a set of GerritChange objects
+    """
+
+    @staticmethod
+    def _query_gerrit(query):
+        query_url = GERRIT_URL + query
+        print("Running query %s" % query_url)
+        response = requests.get(query_url, verify=GERRIT_VERIFY)
+        print("Got response: %s" % response)
+        return json.loads(response.text.replace(")]}'", ''))
+
+    def __init__(self, *query_args):
+        query = '+'.join(query_args)
+        self._get_changes(query)
+
+        # Create GerritChange object corresponding to the query results
+        self._load_changes()
+
+    def _get_changes(self, query):
+        """
+        Get a set of change objects matching the supplied query
+        """
+        query = '/changes/?q={}'.format(query)
+        query += '&o=CURRENT_REVISION&o=CURRENT_COMMIT'
+        response_json = self._query_gerrit(query)
+        self._change_objects = response_json
+
+    def _load_changes(self):
+        self._changes = [GerritChange(change_object=_change_object)
```

Comment:
This is an expensive function. We get some data back from the general gerrit query that might in some cases be enough to determine the questions being asked. It'd be a complicated optimisation to utilise that as opposed to the helper functions in determining dependencies, so I'm happy with this approach. A comment, however, that this /could/ be optimised if needed would be useful.

Reply:
done